### ISSUE TITLE
Improving 3D annotations

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#1351]
## 🚀 Feature
There is a lot of interest these days in better 3D annotation tools, see https://forum.image.sc/t/comparaison-of-some-tools-for-3d-dense-ground-truth-annotations/38918 and https://www.sciencedirect.com/science/article/pii/S2001037020302853 and comments from neubias https://twitter.com/martinjones78/status/1235865025157328896.

We should think about basic improvements to the napari labels (and possibly shapes) layer to improve 3D annotation. If some of them start becoming too domain specific they could be put in plugins, but still good to collect ideas.

A couple things we're missing that come to mind right away is painting into orthoviews. Another could be more "magic wand" like functionality

**Comment by kevinyamauchi** [comment]
Yes! I think another nice feature would be some sort of function to interpolate between labeled slices feature so that people don't have to label every slices if they sampled densely in the out of plane direction (I think somebody mentioned this in the neubias webinar).

**Comment by JoOkuma** [comment]
I might be able to help with this feature.

I've developed some tools using napari with algorithms similar to the magic-wand ([grabber](https://github.com/LIDS-UNICAMP/grabber) and [live-wire](https://github.com/PyIFT/livewire-gui)). However, I don't think they extend very nicely to 3D because the annotation must be done on each slice and then propagated to the volume with some post-processing.

From my limited experience, I prefer flood filling methods for 3D image annotation because they work similarly from 2D to nD images. You just have to extend the image graph adjacency.

The [IFT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.watershed_ift.html) algorithm of Scipy is excellent. I haven't taken a closer look at this implementation. Still, the algorithm can run in O(n) if the image is represented by integers, where n is the number of voxels. It works with an arbitrary number of labels and it can be implemented to compute the algorithm incrementally, re-computing the labels where necessary as the annotation progresses.

I'm not entirely sure if it would be best to add this directly to napari or as a plugin for image segmentation/annotation.

**Comment by sofroniewn** [comment]
> I've developed some tools using napari with algorithms similar to the magic-wand (grabber and live-wire). However, I don't think they extend very nicely to 3D because the annotation must be done on each slice and then propagated to the volume with some post-processing.

Wow! Very cool @JoOkuma those two repos are awesome! We're working on some PRs now to open up our plugin interface, see #1849 and #1816, and it would be great to see how we can turn both of those repos into plugins - cc @tlambert03 @jni.

Looks like the approach you've taken is to define a [custom layer type](https://github.com/LIDS-UNICAMP/grabber/blob/master/src/mypoints.py#L7) which is a subclass of `Points`

**Comment by jni** [comment]
> The IFT algorithm of Scipy is excellent

Note that scikit-image also has an n-dimensional flood fill: https://scikit-image.org/docs/dev/auto_examples/segmentation/plot_floodfill.html as well as n-dimensional watershed.

> I've developed some tools using napari with algorithms similar to the magic-wand (grabber and live-wire).

Those gifs are nothing short of spectacular. 😍😍😍😍!

**Comment by JoOkuma** [comment]
Thanks. I will keep an eye on the plugin interface development to work on it when it is available.

**Comment by kevinyamauchi** [comment]
Hey, @JoOkuma ! The `grabber` and `live-wire` examples look awesome! Thank you for sharing them.
